src/cargo/ops/registry.rs

```diff
@@ -71,9 +71,9 @@ impl RegistryCredentialConfig {
     pub fn is_asymmetric_key(&self) -> bool {
         matches!(self, Self::AsymmetricKey(..))
     }
-    pub fn as_token(&self) -> Option<&str> {
+    pub fn as_token(&self) -> Option<Secret<&str>> {
```

Comment:
Should this be `Secret<&str>` or `&Secret<String>`?
(same for the other `Secret<&str>`)

Reply:
Hmm, I think I like `Secret<&str>` better, although of course I'm open to changing it. I went with `Option<Secret<&str>>` here since it was `Option<&str>` before. More generally, I guess I was thinking of `Secret<T>` as a wrapper similar to `Option<T>`, where I feel like I've seen `Option<&str>` a lot more than `&Option<String>` in fn signatures and such. Though to be honest I'm not sure I really understand the difference between those two types at a deep level. :sweat_smile: 

If you think that `Option<&Secret<String>>` is better, we might also want to consider changing the functions that currently take `Option<Secret<&str>>` as an argument to instead take `Option<&Secret<String>>`.

Reply:
Sorry for not being clear. The `Option` is definitely required here. The question is whether we spell the borrowed form of `Secret<String>` as `&Secret<String>` or if we spell it `Secret<&str>`. As you point out, they're not significantly different types. Whatever we decide should be done consistently.

Reply:
To both issues, it is hard to know the right level of "generic" until you've tried writing the code. Which you have done and I have not. I would appreciate if you would try `&Secret<String>`, I suspect that it will reduce some noise around these conversions and allow us to remove `Secret::as_deref`, but I don't know what other problems it may cause. I will then trust your judgment. After trying each, which one seems better to you?

Reply:
Maybe we could make `map` more useful like:

```rust
    pub fn map<U, F>(&self, f: F) -> Secret<U>
    where
        F: FnOnce(&T) -> U,
    {
        Secret::from(f(&self.inner))
    }
```

Although maybe map isn't the right name for that new signature.

But either way is clearly fine. Whichever you prefer.

Reply:
Let's go with `Secret<&str>` for now. Although it adds a bit of noise around conversions, I feel that it's a little simpler just by virtue of using similar patterns to other wrapper types like `Option` and `Result`.